src/sqlalchemy_oso_cloud/oso.py

```
@@ -101,6 +107,57 @@ def gen_remote_relation_binding(attribute: ColumnProperty, mapper: Mapper, id_co
     }
   }
 
+def _get_classname_by_table(table: Table, registry: registry) -> Optional[str]:
+  for m in registry.mappers:
+    if m.local_table == table:
+      return m.class_.__name__
+  return None
+
+def _get_oso_type(column: NamedColumn, registry: registry, info_key: str, role_mapping_type: str) -> str:
+  foreign_keys = list(column.foreign_keys)
+  oso_type: Optional[str] = None
+  if len(foreign_keys) == 0:
+    if column.info[info_key] is None:
+      raise ValueError(f"Oso role mapping {role_mapping_type} type must be provided if it is not a foreign key")
+    oso_type = column.info[info_key]
+  else:
+    fk = foreign_keys[0].column
+    oso_type = _get_classname_by_table(fk.table, registry)
```

nit: this seems like it allows for a confusing case where the user supplies a custom `actor_type="Something"` on a column that has a foreign key (perhaps to a non-oso-resource table/model), and we just silently ignore their configuration. Would it make sense to either prefer the user-supplied type or loudly error if there's both a foreign key and configured type?

---

i think it makes sense to loudly error so the behavior is explicit. also kind of a workaround for the positional arguments issue - users can't use both a foreign key and a user-supplied type (kwarg).